optuna/multi_objective/hypervolume/wfg.py

```diff
@@ -0,0 +1,83 @@
+import numpy as np
+
+from optuna.multi_objective.hypervolume import BaseHypervolume
+from optuna.multi_objective.hypervolume import compute_2points_volume
+from optuna.multi_objective.hypervolume import dom
+from optuna.multi_objective.hypervolume import DomRelation
+from optuna.multi_objective.hypervolume import Exact2d
+
+
+class WFG(BaseHypervolume):
+    """Hypervolume calculator for any dimension.
+
+        This class exactly calculates the hypervolume for any dimension by using the WFG algorithm.
+        For detail, see `While, Lyndon, Lucas Bradstreet, and Luigi Barone. "A fast way of
+        calculating exact hypervolumes." Evolutionary Computation, IEEE Transactions on 16.1 (2012)
+        : 86-95.`.
+    """
+
+    def __init__(self) -> None:
+        self._r = np.ndarray(shape=())
```

Comment:
How come the reference point is kept as a member (while the solution set isn't)? Might be cleaner just to make it stateless?

Reply:
Yes, because a solution set can change with each internal private function call, but the reference point is immutable once given to the user.